src/actions/mod.rs

```diff
@@ -197,21 +197,81 @@ impl InitActionContext {
     }
 
     fn build<O: Output>(&self, project_path: &Path, priority: BuildPriority, out: O) {
-        struct BuildNotifier<O: Output> {
+
+        lazy_static! {
+            static ref PROGRESS_ID_COUNTER: AtomicUsize = {
+                AtomicUsize::new(0)
+            };
+        }
+
+        fn new_progress_params(id: String, title: &str) -> ProgressParams {
+            ProgressParams {
+                id,
+                title: Some(title.into()),
+                message: None,
+                percentage: None,
+                done: None,
+            }
+        }
+
+        // notifier of diagnostics after the build
+        struct BuildDiagnosticsNotifier<O: Output> {
             out: O,
             active_build_count: Arc<AtomicUsize>,
+            progress_params: ProgressParams,
```

Comment:
This is always used as a template that's cloned for each message, so it'd be good to either add a doc comment explaining that or storing only relevant progress ID here